src/Microsoft.FeatureManagement/ConfigurationFeatureSettingsProvider.cs

```diff
@@ -2,26 +2,52 @@
 // Licensed under the MIT license.
 //
 using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Primitives;
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading;
 
 namespace Microsoft.FeatureManagement
 {
     /// <summary>
     /// A feature settings provider that pulls settings from the .NET Core <see cref="IConfiguration"/> system.
     /// </summary>
-    sealed class ConfigurationFeatureSettingsProvider : IFeatureSettingsProvider
+    sealed class ConfigurationFeatureSettingsProvider : IFeatureSettingsProvider, IDisposable
     {
         private const string FeatureFiltersSectionName = "EnabledFor";
         private readonly IConfiguration _configuration;
+        private readonly ConcurrentDictionary<string, IFeatureSettings> _settings;
+        private readonly IDisposable _changeSubscription;
+        private int _stale = 0;
 
         public ConfigurationFeatureSettingsProvider(IConfiguration configuration)
         {
             _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _settings = new ConcurrentDictionary<string, IFeatureSettings>();
+
+            _changeSubscription = ChangeToken.OnChange(
+                () => _configuration.GetReloadToken(),
+                () => _stale = 1);
+        }
+
+        public void Dispose()
+        {
+            _changeSubscription.Dispose();
         }
 
         public IFeatureSettings TryGetFeatureSettings(string featureName)
+        {
+            if (Interlocked.Exchange(ref _stale, 0) != 0)
+            {
+                _settings.Clear();
+            }
+
+            return _settings.GetOrAdd(featureName, (name) => ReadFeatureSettings(name));
```

Comment:
>ReadFeatureSettings [](start = 61, length = 19)

I through we should just cache our settings not their result? Since we don't know how user implement their filters or how they want to cache it?

Reply:
You're right. This is only caching the settings for the feature, not the result of their evaluations. The state of the feature is evaluated in `FeatureManager`. That class calls on this to get settings for a feature to know what filters to run, etc...

Whenever the configuration underneath us changes we update our cache. This is tracked by the `_stale` bit.

---
In reply to: [350426030](https://github.com/microsoft/FeatureManagement-Dotnet/pull/20#discussion_r350426030) [](ancestors = 350426030)